### PR TITLE
Server render

### PR DESCRIPTION
**Pull request by edpaget**
This adds server-side rendering and handles compatibility issues with include npm packages. I'd like to merge it now to make sure we can keep it current as things change in PFE, but we can continue deploying and using the code the same way.

### CODE REVIEW

**Comment by edpaget** [comment]
This is deployed http://demo.zooniverse.org/panoptes-front-end/server-render/#/

File: app/api/sugar.coffee [hunk 1]
```diff
@@ -1,34 +1,35 @@
 JSONAPIClient = {Resource} = require 'json-api-client'
 authClient = require './auth'
 config = require './config'
-
 SugarClient = require 'sugar-client'
-SugarClient.Primus = require 'sugar-client/primus'
-SugarClient.host = config.sugarHost
 
-window.sugarApiClient = new JSONAPIClient config.sugarHost,
```
Comment (brian-c) [inline app/api/sugar.coffee]: I'm getting a `sugarApiClient is not defined` error when I hit Talk. @parrish does something expect these to be global?

Comment (edpaget) [inline app/api/sugar.coffee]: Ah my bad I can add the assignment to window back in below.

Comment (parrish) [inline app/api/sugar.coffee]: Oops, probably.  It's requirable with `require('<whatever>/api/sugar').api`  Same thing with the socket client.

File: app/classifier/tasks/generic-editor.cjsx [hunk 2]
```diff
@@ -41,7 +41,7 @@ module.exports = React.createClass
             <br />
             <textarea  name="#{@props.taskPrefix}.help" value={@props.task.help ? ""} rows="7" className="standard-input full" onChange={handleChange} />
           </AutoSave>
-          <small className="form-help">Add text and images for a window that pops up when volunteers click “Need some help?” You can use markdown to format this text and add images. The help text can be as long as you need, but you should try to keep it simple and avoid jargon.</small>
+          <small className="form-help">Add text and images for a window? that pops up when volunteers click “Need some help?” You can use markdown to format this text and add images. The help text can be as long as you need, but you should try to keep it simple and avoid jargon.</small>
```
Comment (brian-c) [inline app/classifier/tasks/generic-editor.cjsx]: Stray "?".

File: app/components/dropdown-form.cjsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
+React = require 'react'
```
Comment (brian-c) [inline app/components/dropdown-form.cjsx]: This file has been removed since you forked.

File: app/components/image-selector.cjsx [hunk 4]
```diff
@@ -1,6 +1,6 @@
 React = require 'react'
 LoadingIndicator = require './loading-indicator'
-toBlob = require 'data-uri-to-blob'
+toBlob = require 'data-uri-to-blob' if window.navigator?
```
Comment (brian-c) [inline app/components/image-selector.cjsx]: Should this be `window?.navigator?`?

File: app/components/tooltip.cjsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
 React = require 'react'
-Tether = require 'tether/tether'
+Tether = require 'tether/tether' if window.document?
```
Comment (brian-c) [inline app/components/tooltip.cjsx]: `window?.document?`?

File: app/components/tooltip.cjsx [hunk 6]
```diff
@@ -49,7 +50,7 @@ module.exports = React.createClass
     offset: @props.offset
     targetOffset: @props.targetOffset
     constraints: [
-      {to: 'window', pin: @props.pin}
+      {to: 'window?', pin: @props.pin}
```
Comment (brian-c) [inline app/components/tooltip.cjsx]: Extra "?" in this string.

File: app/pages/lab/how-to-page.cjsx [hunk 7]
```diff
@@ -136,7 +136,7 @@ counterpart.registerTranslations 'en',
 
       You can use markdown in the main text.
 
-      _Help Text_: Add text and images for a pop-up help window. This is shown next to the main text of the task in the main classification interface, when the volunteer clicks a button asking for help. You can use markdown in this text, and link to other images to help illustrate your description. The help text can be as long as you need, but you should try to keep it simple and avoid jargon. One thing that is useful in the help text is a concise description of why you are asking for this particular information.
+      _Help Text_: Add text and images for a pop-up help window?. This is shown next to the main text of the task in the main classification interface, when the volunteer clicks a button asking for help. You can use markdown in this text, and link to other images to help illustrate your description. The help text can be as long as you need, but you should try to keep it simple and avoid jargon. One thing that is useful in the help text is a concise description of why you are asking for this particular information.
```
Comment (brian-c) [inline app/pages/lab/how-to-page.cjsx]: Extra "?".

File: app/pages/project/classify.cjsx [hunk 8]
```diff
@@ -215,12 +215,12 @@ module.exports = React.createClass
...
     # It's not perfect, but it should make the location of everything more obvious.
-    lineHeight = parseFloat getComputedStyle(document.body).lineHeight
+    lineHeight = parseFloat getComputedStyle(document?.body).lineHeight
     el = @getDOMNode()
     space = (innerHeight - el.offsetHeight) / 2
     idealScrollY = el.offsetTop - space
-    if Math.abs(idealScrollY - scrollY) > lineHeight
-      animatedScrollTo document.body, el.offsetTop - space, 333
+    #if Math.abs(idealScrollY - scrollY) > lineHeight
+      #animatedScrollTo document?.body, el.offsetTop - space, 333
```
Comment (brian-c) [inline app/pages/project/classify.cjsx]: I'm all for removing this feature, but not by commenting it out (and probably not in this PR).

File: app/talk/comment.cjsx [hunk 9]
```diff
@@ -29,7 +29,7 @@ module?.exports = React.createClass
     onDeleteComment: React.PropTypes.func # passed (commentId) on click
     onLikeComment: React.PropTypes.func # passed (commentId) on like
     onClickReply: React.PropTypes.func # passed (user, comment) on click
-    active: React.PropTypes.bool  # optional active switch: scroll window to comment and apply styling
+    active: React.PropTypes.bool  # optional active switch: scroll window? to comment and apply styling
```
Comment (brian-c) [inline app/talk/comment.cjsx]: Extra "?".

File: app/talk/sticky-discussion-list.cjsx [hunk 10]
```diff
@@ -1,8 +1,6 @@
+React = require 'react'
 talkClient = require '../api/talk'
 
-placeholder = document.createElement 'li'
```
Comment (brian-c) [inline app/talk/sticky-discussion-list.cjsx]: `placeholder` is still referenced throughout the file.

File: app/components/image-selector.cjsx (hunk 4 above)
Comment (edpaget) [inline app/components/image-selector.cjsx]: Yeah I added a window-shim, so the actual test for things that won't load at all in the server environment ends up being `if window.navigator?`

File: start.js [hunk 11]
```diff
@@ -0,0 +1,6 @@
+require('coffee-react/register');
+require('babel/register')({
```
Comment (brian-c) [inline start.js]: What's babel doing here? Doesn't look like server.js needs it.

Comment (edpaget) [inline start.js]: markdownz does.

File: app/talk/sticky-discussion-list.cjsx (hunk 10 above)
Comment (edpaget) [inline app/talk/sticky-discussion-list.cjsx]: I've added a replacement for this, but sticky reordering doesn't seem to work for me at all in Firefox or Chrome on the main site anyway.

**Comment by brian-c** [comment]
Sorry @edpaget, one last thing, I think: This is missing json-api-client from package.json. Not sure what version you're on, but 2.0.0 doesn't seem to be working.

File: app/pages/project/classify.cjsx (hunk 8 above)
Comment (edpaget) [inline app/pages/project/classify.cjsx]: Ah this may be fine now. I'll check back into into it.

**Comment by edpaget** [comment]
Whoops I must have lost it when I was dealing the API Client revert changes.

**Comment by edpaget** [comment]
The API Client should be restored and the scroll to point has been added.

**Comment by brian-c** [comment]
The `animatedScrollTo = require 'animated-scrollto'` is still commented out in pages/classify.cjsx.

**Comment by edpaget** [comment]
Fixed.

**Comment by brian-c** [comment]
Looks good now! Thanks for all that.

**Pull request merged**